moonraker/components/sensor.py

```diff
@@ -0,0 +1,315 @@
+# Generic sensor support
+#
+# Copyright (C) 2022 Morton Jonuschat <mjonuschat+moonraker@gmail.com>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+# Component to read additional generic sensor data and make it
+# available to clients
+from __future__ import annotations
+
+import logging
+from collections import defaultdict, deque
+from dataclasses import dataclass, replace
+from functools import partial
+
+# Annotation imports
+from typing import (
+    Any,
+    DefaultDict,
+    Deque,
+    Dict,
+    List,
+    Optional,
+    Type,
+    TYPE_CHECKING,
+    Union,
+)
+
+if TYPE_CHECKING:
+    from confighelper import ConfigHelper
+    from websockets import WebRequest
+
+SENSOR_UPDATE_TIME = 1.0
+
+
+@dataclass(frozen=True)
+class SensorConfiguration:
+    id: str
+    name: str
+    type: str
+    source: str = ""
+
+
+if TYPE_CHECKING:
+    from confighelper import ConfigHelper
+
+    from .mqtt import MQTTClient
+
+
+def _set_result(
+    name: str, value: Union[int, float], store: Dict[str, Union[int, float]]
+) -> None:
+    if not isinstance(value, (int, float)):
+        store[name] = float(value)
+    else:
+        store[name] = value
+
+
+@dataclass(frozen=True)
+class Sensor:
+    config: SensorConfiguration
+    values: Dict[str, Deque[Union[int, float]]]
+
+
+class BaseSensor:
+    def __init__(self, name: str, cfg: ConfigHelper, store_size: int = 1200) -> None:
+        self.server = cfg.get_server()
+        self.error_state: Optional[str] = None
+
+        self.config = SensorConfiguration(
+            id=name,
+            type=cfg.get("type"),
+            name=cfg.get("name", name),
+        )
+        self.last_measurements: Dict[str, Union[int, float]] = {}
+        self.last_value: Dict[str, Union[int, float]] = {}
+        self.values: DefaultDict[str, Deque[Union[int, float]]] = defaultdict(
+            lambda: deque(maxlen=store_size)
+        )
+
+    def _update_sensor_value(self, eventtime: float) -> float:
+        """
+        Append the last updated value to the store.
+        """
+        for key, value in self.last_measurements.items():
+            self.values[key].append(value)
+
+        # Copy the last measurements data
+        self.last_value = {**self.last_measurements}
+
+        return eventtime + SENSOR_UPDATE_TIME
```

Comment:
It isn't necessary to return a value here since the timer is now in the main component.